### PR TITLE
Refactor `lock` method

### PR DESCRIPTION
**Pull request by mattverse**
Sub component of: #1838 

## What is the purpose of the change

The lock method in the lockup module was a public method before current change, which was an imo dangerous design, as it is and should be only called by either of the two methods
- `CreateLock`
- `AddTokensToLockByID`

That being said, there shouldn't be any cases where the `Lock` method is called directly, but called only through the two entrypoints mentioned above, not to mention the code duplication that we had for the two methods above. 

This PR changes the Lock method into a private method and refactors code duplication logic as well.

## Brief Changelog
- API change for the `Lock` method in x/lockup module.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not documented

### CODE REVIEW

File: x/lockup/keeper/lock.go [hunk 1]
```diff
@@ -66,100 +66,76 @@ func (k Keeper) AddToExistingLock(ctx sdk.Context, owner sdk.AccAddress, coin sd
...
-	if err != nil {
-		return nil, err
+	for _, synthlock := range k.GetAllSyntheticLockupsByLockup(ctx, lock.ID) {
+		k.accumulationStore(ctx, synthlock.SynthDenom).Increase(accumulationKey(synthlock.Duration), tokensToAdd.Amount)
 	}
 
 	if k.hooks == nil {
 		return lock, nil
 	}
+	k.hooks.AfterAddTokensToLock(ctx, lock.OwnerAddress(), lock.GetID(), sdk.NewCoins(tokensToAdd))
```
Comment (alexanderbez) [inline x/lockup/keeper/lock.go:92]:
```suggestion

	k.hooks.AfterAddTokensToLock(ctx, lock.OwnerAddress(), lock.GetID(), sdk.NewCoins(tokensToAdd))
```

**Review by alexanderbez — APPROVED** [review]

File: x/lockup/keeper/lock.go [hunk 2]
```diff
@@ -66,100 +66,76 @@ func (k Keeper) AddToExistingLock(ctx sdk.Context, owner sdk.AccAddress, coin sd
 // Tokens locked are sent and kept in the module account.
 // This method alters the lock state in store, thus we do a sanity check to ensure
 // lock owner matches the given owner.
-func (k Keeper) AddTokensToLockByID(ctx sdk.Context, lockID uint64, owner sdk.AccAddress, coin sdk.Coin) (*types.PeriodLock, error) {
+func (k Keeper) AddTokensToLockByID(ctx sdk.Context, lockID uint64, owner sdk.AccAddress, tokensToAdd sdk.Coin) (*types.PeriodLock, error) {
 	lock, err := k.GetLockByID(ctx, lockID)
+	if err != nil {
+		return nil, err
+	}
```
Comment (ValarDragon) [inline x/lockup/keeper/lock.go:73]: huh, why did the `errcheck` linter not catch this before. Its enabled?

File: x/lockup/keeper/lock.go [hunk 3]
```diff
@@ -66,100 +66,76 @@ func (k Keeper) AddToExistingLock(ctx sdk.Context, owner sdk.AccAddress, coin sd
...
 	if err != nil {
 		return nil, err
 	}
-	if err := k.bk.SendCoinsFromAccountToModule(ctx, lock.OwnerAddress(), types.ModuleName, sdk.NewCoins(coin)); err != nil {
-		return nil, err
-	}
 
-	err = k.addTokenToLock(ctx, lock, coin)
-	if err != nil {
-		return nil, err
```
Comment (ValarDragon) [inline x/lockup/keeper/lock.go:85]: This looks wrong?  addTokenToLock did more, e.g. handling of the main accumulation store? Why was this deleted?

Comment (ValarDragon) [reply x/lockup/keeper/lock.go:85]: Oh the claim is that all functionality is there between the synth lock line, and lock.

Why not keep `addTokenToLock` and just have that function as a subcomponent call `lock`? (But retain the synth lock functionality)

That would be maintaining the standard in this package of having the exported method be doing validation logic & hook / return type logic, and the private method doing the state change

File: x/lockup/keeper/lock.go [hunk 4]
```diff
@@ -66,100 +66,76 @@ func (k Keeper) AddToExistingLock(ctx sdk.Context, owner sdk.AccAddress, coin sd
...
-	store := ctx.KVStore(k.storeKey)
-	bz, err := proto.Marshal(&lock)
-	if err != nil {
-		return err
-	}
-	store.Set(lockStoreKey(lock.ID), bz)
-
-	// add lock refs into not unlocking queue
-	err = k.addLockRefs(ctx, lock)
+	err = k.setLock(ctx, lock)
```
Comment (ValarDragon) [inline x/lockup/keeper/lock.go:132]: ty!

File: x/lockup/keeper/lock.go (hunk 2 above)
Comment (p0mvn) [reply x/lockup/keeper/lock.go:73]: `golangci-lint linters` shows it as enabled. Maybe a bug in the linter?

File: x/lockup/keeper/lock_test.go [hunk 5]
```diff
@@ -144,47 +144,17 @@ func (suite *KeeperTestSuite) TestUnlockPeriodLockByID() {
 	suite.Require().Len(locks, 0)
 }
 
-func (suite *KeeperTestSuite) TestLock() {
```
Comment (p0mvn) [inline x/lockup/keeper/lock_test.go:147]: Does it make sense to do either or both:
-  test `CreateLock`
- use the [`export_test.go`](https://medium.com/@robiplus/golang-trick-export-for-test-aa16cbd7b8cd) trick that will allow us to
    * export `lock` for tests only and
    * keep testing the now unexported `lock` method

File: x/lockup/keeper/lock.go [hunk 6]
```diff
@@ -66,100 +66,76 @@ func (k Keeper) AddToExistingLock(ctx sdk.Context, owner sdk.AccAddress, coin sd
...
+	if err != nil {
+		return nil, err
+	}
 
 	if lock.GetOwner() != owner.String() {
 		return nil, types.ErrNotLockOwner
 	}
 
+	lock.Coins = lock.Coins.Add(tokensToAdd)
+	err = k.lock(ctx, *lock, sdk.NewCoins(tokensToAdd))
```
Comment (AlpinYukseloglu) [inline x/lockup/keeper/lock.go:80]: Out of curiosity, why do we use `sdk.NewCoins(tokensToAdd)` here instead of directly passing `tokensToAdd` in like the `Lock` function does?

File: x/lockup/keeper/lock.go (hunk 3 above)
Comment (AlpinYukseloglu) [reply x/lockup/keeper/lock.go:85]: I still don't fully follow the logic here. There are two calls that `addTokenToLock` makes that are not related to the synth lock related accumulation store update, which are:
1. `setLock` to update state with the new lock instance:
https://github.com/osmosis-labs/osmosis/blob/8f018a6c638b64912c1b22c7df04f84abe667ba0/x/lockup/keeper/lock.go#L100
2. Update to `accumulationStore`:
https://github.com/osmosis-labs/osmosis/blob/8f018a6c638b64912c1b22c7df04f84abe667ba0/x/lockup/keeper/lock.go#L106

How/where are these being accounted for in the new implementation?

File: x/lockup/keeper/lock.go [hunk 7]
```diff
@@ -66,100 +66,76 @@ func (k Keeper) AddToExistingLock(ctx sdk.Context, owner sdk.AccAddress, coin sd
...
 	if err != nil {
 		return nil, err
 	}
-	if err := k.bk.SendCoinsFromAccountToModule(ctx, lock.OwnerAddress(), types.ModuleName, sdk.NewCoins(coin)); err != nil {
-		return nil, err
-	}
```
Comment (AlpinYukseloglu) [inline x/lockup/keeper/lock.go:81]: Just to clarify, this was removed because we're now calling `lock` above, which already sends new locked tokens to the module account?

File: x/lockup/keeper/lock.go (hunk 3 above)
Comment (AlpinYukseloglu) [reply x/lockup/keeper/lock.go:85]: Oops, I see the `lock` call above now. The change that adds an addition parameter to `lock` also makes sense now. Clever refactor :)

File: x/lockup/keeper/lock.go [hunk 8]
```diff
@@ -66,100 +66,76 @@ func (k Keeper) AddToExistingLock(ctx sdk.Context, owner sdk.AccAddress, coin sd
...
 
-	err = k.addTokenToLock(ctx, lock, coin)
-	if err != nil {
-		return nil, err
+	for _, synthlock := range k.GetAllSyntheticLockupsByLockup(ctx, lock.ID) {
+		k.accumulationStore(ctx, synthlock.SynthDenom).Increase(accumulationKey(synthlock.Duration), tokensToAdd.Amount)
 	}
```
Comment (AlpinYukseloglu) [inline x/lockup/keeper/lock.go:87]: Is it okay that we're no longer doing the single coin check that was originally in `addTokenToLock`? Is it checked elsewhere or was it not necessary to begin with?
https://github.com/osmosis-labs/osmosis/blob/8f018a6c638b64912c1b22c7df04f84abe667ba0/x/lockup/keeper/lock.go#L108-L111

**Review by AlpinYukseloglu — COMMENTED** [review]
Left a few clarifying questions but overall LGTM! Love this refactor, cleans things up quite a bit :)

File: x/lockup/keeper/lock.go (hunk 2 above)
Comment (mattverse) [reply x/lockup/keeper/lock.go:73]: oh its just because we weren't using the `lock` method previosuly!

File: x/lockup/keeper/lock.go (hunk 6 above)
Comment (mattverse) [reply x/lockup/keeper/lock.go:80]: we're using a pointer reference here, editing the lock directly makes it impossible to track the newly added tokens to the lock, which is needed for accumulation store, lock refs etc!

File: x/lockup/keeper/lock.go (hunk 7 above)
Comment (mattverse) [reply x/lockup/keeper/lock.go:81]: yeah!

File: x/lockup/keeper/lock.go (hunk 3 above)
Comment (mattverse) [reply x/lockup/keeper/lock.go:85]: @ValarDragon  wdym by ?

> (But retain the synth lock functionality)

File: x/lockup/keeper/lock.go (hunk 8 above)
Comment (mattverse) [reply x/lockup/keeper/lock.go:87]: this has been checked in the ValidateBasic in the msg server!

File: x/lockup/keeper/lock_test.go (hunk 5 above)
Comment (mattverse) [reply x/lockup/keeper/lock_test.go:147]: Ah missed the test for `CreateLock` thanks for catching this! Just added this in the most recent commit!

**Review by AlpinYukseloglu — APPROVED** [review]

**Comment by mattverse** [comment]
merging this upon being blocker for future work for adding testing for lockup!

**Comment by mattverse** [comment]
Apologies if I went ahead and rushed into merging this, I supposed there weren't extra reviews.

Is there something you're concerned about or would like to give a review on? We can either revert this commit or create a new PR depending on the review

**Comment by mattverse** [comment]
A Note to reviewers to clarify no logic has been removed / changed per this PR

**Previous logic for `AddTokensToLockByID`**
- SendCoinsFromAccountToModule
- addTokenToLock
     - lock.Coins = lock.Coins.Add(coin)
     - setLock
     - accumulation store increase
     - synth denom accumulation store increase
     - hooks.AfterAddTokensToLock
- hooks.OnTokenLocked

**Refactored Logic for `AddTokensToLockByID`**
- lock.Coins = lock.Coins.Add(coin)
- lock
   - SendCoinsFromAccountToModule
   - setLock
   - accumulation store increase
   - hooks.OnTokenLocked 
 - synth denom accumulation store increase
 -  hooks.AfterAddTokensToLock
 
---------
 **Previous logic for `CreateLock`**
 - Lock
   - SendCoinsFromAccountToModule
   - store.Set(lock)
   - addlockRefs
   - accumulation store increase
   - hooks.OnTokenLocked
  - setLastLockID
  
  **Refactored Logic for `CreateLock`**
  - lock 
      - SendCoinsFromAccountToModule
     - setLock
     - accumulation store increase
      - hooks.OnTokenLocked  
   - addLockRefs
   - SetLastLockID
